File: matura.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

from math import ceil, floor

def sklej(n):
  if n == 1:
    return 1
  elif n % 2 == 0:
    return n - 1 + 2*sklej(n/2)
  else:
    return n - 1 + sklej((n-1)/2) + sklej((n+1)/2)

def sklej_gen(n):
  tab = [1,1]
  for i in range(2, n + 1):
# I need to use some ugly hacks to make it usable with Python 2 and 3
    tab.append(i - 1 + tab[int(ceil(i/2.0))] + tab[int(floor(i/2.0))])
  return tab

for i in sklej_gen(10000):
  print(i)

File: matura.rb
#!/usr/bin/env ruby
# encoding: UTF-8

def sklej(n)
  if n == 1
    return 1
  elsif n.even?
    return n - 1 + 2*sklej(n/2)
  else
    return n - 1 + sklej((n-1)/2) + sklej((n+1)/2)
  end
end

def sklej_gen(size)
  tab = [1,1]
  (2..size).each do |i|
    tab[i] = i - 1 + tab[(i/2.0).floor] + tab[(i/2.0).ceil]
  end
  tab
end

#(1..10000).each { |i| puts sklej(i) }
sklej_gen(10000).each { |i| puts i }

